vendor/Ltac2/g_ltac2.mlg

```diff
@@ -0,0 +1,933 @@
+        { StrTyp (isrec, l) }
+    ] ]
+  ;
+  tac2def_ext:
+    [ [ "@"; IDENT "external"; id = locident; ":"; t = tac2type LEVEL "5"; ":=";
+        plugin = Prim.string; name = Prim.string ->
+      { let ml = { mltac_plugin = plugin; mltac_tactic = name } in
+        StrPrm (id, t, ml) }
+    ] ]
+  ;
+  syn_node:
+    [ [ "_" -> { CAst.make ~loc None }
+      | id = Prim.ident -> { CAst.make ~loc (Some id) }
+    ] ]
+  ;
+  sexpr:
+    [ [ s = Prim.string -> { SexprStr (CAst.make ~loc s) }
+      | n = Prim.integer -> { SexprInt (CAst.make ~loc n) }
+      | id = syn_node -> { SexprRec (loc, id, []) }
+      | id = syn_node; "("; tok = LIST1 sexpr SEP "," ; ")" ->
+        { SexprRec (loc, id, tok) }
+    ] ]
+  ;
+  syn_level:
+    [ [ -> { None }
+      | ":"; n = Prim.integer -> { Some n }
+    ] ]
+  ;
+  tac2def_syn:
+    [ [ "Notation"; toks = LIST1 sexpr; n = syn_level; ":=";
+        e = tac2expr ->
+        { StrSyn (toks, n, e) }
+    ] ]
+  ;
+  lident:
+    [ [ id = Prim.ident -> { CAst.make ~loc id } ] ]
+  ;
+  globref:
+    [ [ "&"; id = Prim.ident -> { CAst.make ~loc (QHypothesis id) }
+      | qid = Prim.qualid -> { CAst.make ~loc @@ QReference qid }
+    ] ]
+  ;
+END
+
+(* Quotation scopes used by notations *)
+
+{
+
+open Tac2entries.Pltac
+
+let loc_of_ne_list l = Loc.merge_opt (fst (List.hd l)) (fst (List.last l))
+
+}
+
+GRAMMAR EXTEND Gram
+  GLOBAL: q_ident q_bindings q_intropattern q_intropatterns q_induction_clause
+          q_conversion q_rewriting q_clause q_dispatch q_occurrences q_strategy_flag
+          q_destruction_arg q_reference q_with_bindings q_constr_matching
+          q_goal_matching q_hintdb q_move_location q_pose q_assert;
+  anti:
+    [ [ "$"; id = Prim.ident -> { QAnti (CAst.make ~loc id) } ] ]
+  ;
+  ident_or_anti:
+    [ [ id = lident -> { QExpr id }
+      | "$"; id = Prim.ident -> { QAnti (CAst.make ~loc id) }
+    ] ]
+  ;
+  lident:
+    [ [ id = Prim.ident -> { CAst.make ~loc id } ] ]
+  ;
+  lnatural:
+    [ [ n = Prim.natural -> { CAst.make ~loc n } ] ]
+  ;
+  q_ident:
+    [ [ id = ident_or_anti -> { id } ] ]
+  ;
+  qhyp:
+    [ [ x = anti -> { x }
+      | n = lnatural -> { QExpr (CAst.make ~loc @@ QAnonHyp n) }
+      | id = lident -> { QExpr (CAst.make ~loc @@ QNamedHyp id) }
+    ] ]
+  ;
+  simple_binding:
+    [ [ "("; h = qhyp; ":="; c = Constr.lconstr; ")" ->
+        { CAst.make ~loc (h, c) }
+    ] ]
+  ;
+  bindings:
+    [ [ test_lpar_idnum_coloneq; bl = LIST1 simple_binding ->
+        { CAst.make ~loc @@ QExplicitBindings bl }
+      | bl = LIST1 Constr.constr ->
+        { CAst.make ~loc @@ QImplicitBindings bl }
+    ] ]
+  ;
+  q_bindings:
+    [ [ bl = bindings -> { bl } ] ]
+  ;
+  q_with_bindings:
+    [ [ bl = with_bindings -> { bl } ] ]
+  ;
+  intropatterns:
+    [ [ l = LIST0 nonsimple_intropattern -> { CAst.make ~loc l } ] ]
+  ;
+(*   ne_intropatterns: *)
+(*     [ [ l = LIST1 nonsimple_intropattern -> l ]] *)
+(*   ; *)
+  or_and_intropattern:
```

Comment:
There seem to be many productions that duplicate those in existing .mlg files.  Ex: the same productions for`or_and_intropattern` is already in `g_tactic.mlg`.  This should be cleaned up.

The tool in https://github.com/coq/coq/pull/9884 could be very helpful in identifying the duplications.  I hope to see that tool merged fairly soon, perhaps within the next week or two.

Reply:
@ppedrot what should I do about this?

Reply:
For now, nothing I believe. This should be part of a later refinement.

Reply:
Furthermore, in this particular case, the grammars are different. Ltac2 also handles antiquotations by means of `$x` tokens.